### PR TITLE
Link Panel: issues with special characters

### LINKED ISSUE #6482: Link Panel: issues with special characters
## Bug Description

The link title was incorrectly filled in automatically when there are special characters.

https://user-images.githubusercontent.com/66372350/108758901-a6eb4200-7519-11eb-872b-7e91a7030bfb.mov

## Expected Behaviour

Link title should automatically fill in description field with correct special character

image

## Steps to Reproduce

1. Enter a URL to apply a link. URL should contain a link title with a special character like this: https://www.selectgame.com.br/chizuru-kagura-estara-em-the-king-of-fighters-xv-novo-trailer/ 
2. Notice instead of `estará` the description field automatically filled in `estarÃ¡`

## Screenshots
Inserted this [link](https://www.selectgame.com.br/chizuru-kagura-estara-em-the-king-of-fighters-xv-novo-trailer/) that contained special character `Chizuru Kagura estará em` but it automatically displayed the following: 
image

## Additional Context
Related topic: 
https://wordpress.org/support/topic/error-in-linking-with-special-characters-in-title-url/ 

Workaround: Manually edit the link description.

- Plugin Version: 1.4.0
- WordPress Version: 5.6.1
- Operating System: macOS
- Browser: Chrome

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

As documented in the DOMElement class [docs](https://www.php.net/manual/en/class.domelement.php).

> The DOM extension uses UTF-8 encoding. Use utf8_encode() and utf8_decode() to work with texts in ISO-8859-1 encoding or iconv for other encodings.

## Testing Instructions

1. Create a story. 
2. Create an element. 
3. Add this link - https://www.selectgame.com.br/chizuru-kagura-estara-em-the-king-of-fighters-xv-novo-trailer/
4. See that title embed with character correct. 

You could also embed other links from https://www.selectgame.com.br/ to test. 

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testiing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #6482

### CODE REVIEW

File: includes/REST_API/Link_Controller.php [hunk 1]
```diff
@@ -218,6 +222,10 @@ public function parse_link( $request ) {
 			$og_description_query = $xpath->query( '//meta[@property="og:description"]' );
 			$description          = $this->get_dom_attribute_content( $og_description_query, 'content' );
 		}
+		
+		if ( $description ) {
+			$description = utf8_decode( $description );
```
Comment (swissspidy) [inline includes/REST_API/Link_Controller.php]: Hmm can we instead ensure that the whole document is encoded as UTF-8? Would be good to stay in UTF-8 from start to finish.

I think we can easily achieve this by changing `html_to_xpath()` in `Document_Parser` to use the `Document` util from the AMP toolbox.

It's a general utility class for working with `DOMDocument` that is useful outside of AMP context.

Should be as simply as this:

```php
protected function html_to_xpath( $html ) {
	$document = Google\Web_Stories_Dependencies\AmpProject\Dom\Document::fromHtml( $html );
    return $document->xpath;
}
```

We could probably simplify a few more things by using that class.

Comment (spacedmonkey) [reply includes/REST_API/Link_Controller.php]: Oh, I didn't know about this class. Very cool. I changed it here - https://github.com/google/web-stories-wp/pull/6676/commits/ef477690cf7c9dc66699532b1a783ecb32a19397

**Review by swissspidy — APPROVED** [review]
Looks good 👍 

Just needs some QA instructions.

**Comment by swissspidy** [comment]
We might want to consider removing `Document_Parser` trait and instead extending the `Document` class with the required method. Perhaps that method could even be contributed upstream.

**Comment by spacedmonkey** [comment]
> We might want to consider removing Document_Parser trait and instead extending the Document class with the required method. Perhaps that method could even be contributed upstream.

I did think about that, by extending the class. But this fix, fixes the problem, we can refactor later.

**Comment by swissspidy** [comment]
> > We might want to consider removing Document_Parser trait and instead extending the Document class with the required method. Perhaps that method could even be contributed upstream.
> 
> I did think about that, by extending the class. But this fix, fixes the problem, we can refactor later.

Oh yeah totally, didn't mean doing that in this PR

**Comment by spacedmonkey** [comment]
>Just needs some QA instructions.

Added.

**Pull request merged**